Userland/Utilities/pkg/BuildablePort.cpp

```diff
@@ -0,0 +1,109 @@
+            return Error::from_string_view("Input file is missing a hash or git revision"sv);
+        auto fragment = maybe_fragment.release_value();
+
+        if (url.scheme() == "https" || url.scheme() == "http") {
+            auto bigint_hash = Crypto::UnsignedBigInteger::from_base(16, fragment);
+            VERIFY(bigint_hash.length() * Crypto::UnsignedBigInteger::BITS_IN_WORD / 8 == Crypto::Hash::SHA256::digest_size());
+            Crypto::Hash::SHA256::DigestType hash;
+            bigint_hash.export_data({ &hash.data, hash.data_length() });
+            file_list.append({ .url = move(url), .type_specific_info = PortInputFile::HTTPInfo { hash } });
+        } else if (url.scheme() == "git+https") {
+            file_list.append({ .url = move(url), .type_specific_info = PortInputFile::GitInfo { move(fragment) } });
+        } else {
+            return Error::from_string_view("Unsupported input file scheme"sv);
+        }
+    }
+
+    return file_list;
+}
+
+ErrorOr<BuildablePort> BuildablePort::from_available_port(AvailablePort const& port)
+{
+    auto port_root = port.local_port_root();
+    if (!FileSystem::exists(port_root.string()) || !FileSystem::is_directory(port_root.string()))
+        return Error::from_string_view("Port is not available in the local file system"sv);
+
+    auto package_base_path = LexicalPath::join(port_root.string(), "package.sh"sv);
+    if (!FileSystem::exists(package_base_path.string()))
+        return Error::from_string_view("Port is missing a package.sh script, it may be an empty directory"sv);
+
+    auto shell = Shell::Shell::construct(true);
+    {
+        // FIXME: Find a nicer way of disabling Shell's output.
+        auto old_stdout = TRY(Core::System::dup(STDOUT_FILENO));
+        auto old_stderr = TRY(Core::System::dup(STDERR_FILENO));
+        auto null_device = TRY(Core::System::open("/dev/null"sv, O_RDWR));
+        TRY(Core::System::dup2(null_device, STDOUT_FILENO));
+        TRY(Core::System::dup2(null_device, STDERR_FILENO));
+        ScopeGuard restore_stdio { [&] {
+            MUST(Core::System::dup2(old_stdout, STDOUT_FILENO));
+            MUST(Core::System::dup2(old_stderr, STDERR_FILENO));
+        } };
+
+        auto was_successful = shell->run_file(package_base_path.string());
+        if (!was_successful)
+            return Error::from_string_view("Port package.sh couldn't be executed."sv);
+    }
+
+    BuildablePort buildable_port;
+    buildable_port.m_absolute_path = port_root;
+
+    auto port_name = TRY(shell->look_up_local_variable("port"sv));
+    // Do not check for is_string() here, the port name may have one of several string-convertible types.
+    // If the type is actually incompatible, resolve_as_string will catch this.
+    if (port_name == nullptr)
+        return Error::from_string_view("Port name is not a string"sv);
+    buildable_port.m_name = TRY(const_cast<Shell::AST::Value&>(*port_name).resolve_as_string(shell));
```

Comment:
I'd suggest you just let the shell handle the conversion, as the empty string is an invalid port name anyway:
```suggestion
    buildable_port.m_name = TRY(shell->local_variable_or("port", ""));
    if (buildable_port.m_name.is_empty())
        return Error::from_string_view("...");
```